fw/http_parser.c

```diff
@@ -2312,65 +2330,107 @@ __req_parse_cache_control(TfwHttpReq *req, unsigned char *data, size_t len)
 	__FSM_REQUIRE_FIRST_DIGIT(Req_I_CC_MinFreshVBeg, Req_I_CC_MinFreshV);
 
 	__FSM_STATE(Req_I_CC_MinFreshV) {
+		parser->cache_control.filled = true;
+
 		__fsm_sz = __data_remain(p);
 		__fsm_n = parse_uint_list(p, __fsm_sz, &parser->_acc);
 		if (__fsm_n == CSTR_POSTPONE)
 			__msg_hdr_chunk_fixup(data, len);
 		if (__fsm_n < 0)
-			return __fsm_n;
+			__FSM_EXIT(__fsm_n);
 		req->cache_ctl.min_fresh = parser->_acc;
 		req->cache_ctl.flags |= TFW_HTTP_CC_MIN_FRESH;
 		__FSM_I_MOVE_n(Req_I_EoT, __fsm_n);
 	}
 
 	__FSM_STATE(Req_I_CC_MaxStale) {
+		parser->cache_control.filled = true;
+
 		if (c == '=')
 			__FSM_I_MOVE(Req_I_CC_MaxStaleVBeg);
 		if (IS_WS(c) || c == ',' || IS_CRLF(c)) {
 			req->cache_ctl.max_stale = UINT_MAX;
 			req->cache_ctl.flags |= TFW_HTTP_CC_MAX_STALE;
 			__FSM_I_JMP(Req_I_EoT);
 		}
+		parser->cache_control.dir_flag = TFW_HTTP_CC_OTHER;
+		/* something like "max-staledfgh$!dgh" */
 		__FSM_I_JMP(Req_I_CC_Ext);
 	}
 
 	__FSM_REQUIRE_FIRST_DIGIT(Req_I_CC_MaxStaleVBeg, Req_I_CC_MaxStaleV);
 
 	__FSM_STATE(Req_I_CC_MaxStaleV) {
+		parser->cache_control.filled = true;
+
 		__fsm_sz = __data_remain(p);
 		__fsm_n = parse_uint_list(p, __fsm_sz, &parser->_acc);
 		if (__fsm_n == CSTR_POSTPONE)
 			__msg_hdr_chunk_fixup(data, len);
 		if (__fsm_n < 0)
-			return __fsm_n;
+			__FSM_EXIT(__fsm_n);
 		req->cache_ctl.max_stale = parser->_acc;
 		req->cache_ctl.flags |= TFW_HTTP_CC_MAX_STALE;
 		__FSM_I_MOVE_n(Req_I_EoT, __fsm_n);
 	}
 
 	__FSM_STATE(Req_I_CC_Ext) {
-		/* TODO: process cache extensions. */
-		__FSM_I_MATCH_MOVE(qetoken, Req_I_CC_Ext);
-		c = *(p + __fsm_sz);
-		if (IS_WS(c) || c == ',')
-			__FSM_I_MOVE_n(Req_I_EoT, __fsm_sz + 1);
-		if (IS_CRLF(c))
-			return __data_off(p + __fsm_sz);
-		return CSTR_NEQ;
+		/* Any flag we don't understand.
+		 * Here we just skip all the tokens, double quotes and equal signs.
+		 */
+		__FSM_I_MATCH_MOVE_fixup_finish(qetoken, Req_I_CC_Ext, 0, {
+			parser->cache_control.filled = true;
+			if (parser->cache_control.dir_flag == 0) {
+				parser->cache_control.dir_flag = TFW_HTTP_CC_OTHER;
+			}
+		});
+		if (__fsm_sz > 0) {
+			if (parser->cache_control.dir_flag == 0) {
+				parser->cache_control.dir_flag = TFW_HTTP_CC_OTHER;
+			}
+			parser->cache_control.filled = true;
+		}
+
+		__FSM_I_MOVE_n(Req_I_EoT, __fsm_sz);
 	}
 
 	/* End of term. */
 	__FSM_STATE(Req_I_EoT) {
-		if (IS_WS(c) || c == ',')
+		if (IS_WS(c))
 			__FSM_I_MOVE(Req_I_EoT);
 
+		if (c == ',')
+			__FSM_I_MOVE(Req_I_After_Comma);
+		/* Tokens have to be separated by comma */
+		if (IS_TOKEN(c))
+			__FSM_EXIT(TFW_BLOCK);
+
+		if (IS_CRLF(c)) {
+			/* Cache-Control header cannot be empty */
+			if (!parser->cache_control.filled)
+				__FSM_EXIT(TFW_BLOCK);
```

Comment:
+1 for @Dmitry-Gouriev 's comments. `dir_flag` is just a replacement for old `_acc` and I'm good with it. But Dmitry pointed out that probably we can directly set the flags, w/o collecting them in `_acc`. `_acc` stands for "accumulator" and the primary target for the field is to accumulate multi-digit number values, which seem not necessary for flags.

`cache_control.filled` still looks unnatural, but there isn't so many code around it and now FSM looks much more consistent (so I closed #1539). However, probably we can determine whether Cache-Control had a directive with introducing only one new state. In this case we won't need `filled`.

If we can get rid of both the `filled` and `dir_flag`, then we can get rid of `Req_I_CC_start` and, with one new state instead of `filled`, we'll have the same number of states and more straightforward and likely faster FSM.

BTW since `Req_I_CC_start` is executed only once on very first FSM entrance, we can initialize the variable just at the beginning of the function and we don't need a state for it.